lib/diagnostics_channel.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypePush,
+  FunctionPrototypeBind,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      ArrayPrototypePush(this._subscribers, subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        const fn = FunctionPrototypeBind(subscription.onMessage, subscription);
+        ArrayPrototypePush(this._subscribers, fn);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.ifActive = undefined;
+    this.name = name;
+  }
+
+  subscribe(subscription) {
```

Comment:
I think this is a private API, so it should be `_subscribe` (same is valid for `ActivePublisher`).

Reply:
No, it's not intended to be private. You should be able to subscribe directly to a channel, if you have the channel object.

Reply:
In that case it's missing in the docs.
But my feeling was that `Publisher` (Channel) objects are intended to be used only by the publishing side which usually is not subscribing to itself.